[AIRFLOW-719] Prevent DAGs from ending prematurely

dhuang opened the pull request:
DAGs using ALL_SUCCESS and ONE_SUCCESS trigger rules were ending
prematurely when upstream tasks were skipped. Changes mean that the
ALL_SUCCESS and ONE_SUCCESS triggers rule encompasses both SUCCESS and
SKIPPED tasks.

See original PR https://github.com/apache/incubator-airflow/pull/1961 by @alexvanboxel for additional discussion. I applied the same fix to ONE_SUCCESS and added tests on top of his changes.

Please accept this PR that addresses the following issues:
- https://issues.apache.org/jira/browse/AIRFLOW-719

Testing Done:
- Updated parameterized task dependency unit tests with new rules. Also fixed the "done" count on all these tests (there were some mismatches). Added some additional tests that go through the DagRun.update_state() flow.
- Manually tested a few DAGs that used LatestOnly, ShortCircuit, and BranchPython operators that skipped. Also double checked a number of relevant example DAGs.

bolkedebruin commented on tests/dags/test_model_dags.py:
```diff
@@ -0,0 +1,38 @@
+# -*- coding: utf-8 -*-
```
I'm not a big fan of having test DAGs in separate files, especially since the file name does not match the DAG name, making it harder to find out where a DAG lives. Can you rename your file to the dag name and  refer to it inside the tests where you are using it?

dhuang replied:
Done

bolkedebruin commented on tests/models.py:
```diff
@@ -117,13 +118,71 @@ def test_dag_as_context_manager(self):
+            external_trigger=False,
+        )
+
+        if task_states is not None:
```
dag_run.get_task_instances, then you know for sure you are getting the right ones

dhuang replied:
Ah yes that's better, fixed.

dhuang committed:
[AIRFLOW-719] Prevent DAGs from ending prematurely

DAGs using ALL_SUCCESS and ONE_SUCCESS trigger rules were ending
prematurely when upstream tasks were skipped. Changes mean that the
ALL_SUCCESS and ONE_SUCCESS triggers rule encompasses both SUCCESS and
SKIPPED tasks.
